pkg/chain/ethereum/tbtc.go

```diff
@@ -996,3 +1031,148 @@ func (tc *TbtcChain) activeWalletPublicKey() ([]byte, bool, error) {
 
 	return publicKeyBytes, true, nil
 }
+
+func (tc *TbtcChain) OnDepositSweepProposalSubmitted(
+	handler func(event *tbtc.DepositSweepProposalSubmittedEvent),
+) subscription.EventSubscription {
+	onEvent := func(
+		proposal tbtcabi.WalletCoordinatorDepositSweepProposal,
+		proposalSubmitter common.Address,
+		blockNumber uint64,
+	) {
+		handler(&tbtc.DepositSweepProposalSubmittedEvent{
+			Proposal:          convertDepositSweepProposalFromAbiType(proposal),
+			ProposalSubmitter: chain.Address(proposalSubmitter.Hex()),
+			BlockNumber:       blockNumber,
+		})
+	}
+
+	return tc.walletCoordinator.
+		DepositSweepProposalSubmittedEvent(nil, nil).
+		OnEvent(onEvent)
+}
+
+func (tc *TbtcChain) PastDepositSweepProposalSubmittedEvents(
+	filter *tbtc.DepositSweepProposalSubmittedEventFilter,
+) ([]*tbtc.DepositSweepProposalSubmittedEvent, error) {
+	var startBlock uint64
+	var endBlock *uint64
+	var proposalSubmitter []common.Address
+	var walletPublicKeyHash [20]byte
+
+	if filter != nil {
+		startBlock = filter.StartBlock
+		endBlock = filter.EndBlock
+
+		for _, ps := range filter.ProposalSubmitter {
+			proposalSubmitter = append(
+				proposalSubmitter,
+				common.HexToAddress(ps.String()),
+			)
+		}
+
+		walletPublicKeyHash = filter.WalletPublicKeyHash
+	}
+
+	events, err := tc.walletCoordinator.PastDepositSweepProposalSubmittedEvents(
+		startBlock,
+		endBlock,
+		proposalSubmitter,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	convertedEvents := make([]*tbtc.DepositSweepProposalSubmittedEvent, 0)
+	for _, event := range events {
+		// If the wallet PKH filter is set, omit all events that target
+		// different wallets.
+		if walletPublicKeyHash != [20]byte{} {
+			if event.Proposal.WalletPubKeyHash != walletPublicKeyHash {
+				continue
+			}
+		}
```

Comment:
Is it needed at all? The filter - if set - should not return events with wallets with a different key.

It's the same situation as with the proposal submitter filter.

Reply:
The main difference is that `walletPublicKeyHash` is not indexed as this is part of the proposal struct. We need to do that filtering manually and this is what we do here.

Reply:
Ah, I see now. `walletPublicKeyHash` is not passed to `PastDepositSweepProposalSubmittedEvents`. Makes sense.